resolver/dns_resolver.ml

```diff
@@ -288,65 +292,71 @@ let handle_reply t now ts proto sender packet reply =
+                          (fun ds -> not (ds.Ds.digest_type = SHA1))
+                          ds_set
+                      else
+                        ds_set
+                    in
+                    Option.map (fun (_, dnskeys) ->
+                        Rr_map.Ds_set.fold (fun ds acc ->
+                            match Dnssec.validate_ds zone dnskeys ds with
+                            | Ok key -> Rr_map.Dnskey_set.add key acc
+                            | Error `Msg msg ->
+                              Logs.warn (fun m -> m "couldn't validate DS (for %a): %s"
+                                            Domain_name.pp zone msg);
+                              acc)
+                          ds_set Rr_map.Dnskey_set.empty)
+                      keys
+                  | _ ->
+                    Logs.warn (fun m -> m "no DS in cache for %a" Domain_name.pp zone);
+                    None
+                end
               | _ ->
-                Logs.warn (fun m -> m "no DS in cache for %a" Domain_name.pp zone);
-                None
-            end
-          | _ ->
-            let cache, dnskeys = Dns_cache.get t.cache ts zone Rr_map.Dnskey in
-            { t with cache },
-            match dnskeys with
-            | Ok (`Entry (_, dnskey_set), _) -> Some dnskey_set
-            | _ ->
-              Logs.warn (fun m -> m "no DNSKEYS in cache for %a" Domain_name.pp zone);
-              None
-        in
-        let* packet, signed =
-          Option.fold
-            ~none:(Ok (packet, false))
-            ~some:(fun dnskeys ->
-                let* packet =
-                  Result.map_error (fun (`Msg msg) ->
-                      Logs.err (fun m -> m "error %s verifying reply %a"
-                                   msg Packet.pp_reply reply))
-                    (Dnssec.verify_packet now dnskeys packet)
-                in
-                Ok (packet, true))
-            dnskeys
+                let cache, dnskeys = Dns_cache.get t.cache ts zone Rr_map.Dnskey in
+                { t with cache },
+                match dnskeys with
+                | Ok (`Entry (_, dnskey_set), _) -> Some dnskey_set
+                | _ ->
+                  Logs.warn (fun m -> m "no DNSKEYS in cache for %a" Domain_name.pp zone);
+                  None
+            in
+            let* packet, signed =
+              Option.fold
+                ~none:(Ok (packet, false))
+                ~some:(fun dnskeys ->
+                    let* packet =
+                      Result.map_error (fun (`Msg msg) ->
+                          Logs.err (fun m -> m "error %s verifying reply %a"
+                                       msg Packet.pp_reply reply))
+                        (Dnssec.verify_packet now dnskeys packet)
+                    in
+                    Ok (packet, true))
+                dnskeys
+            in
+            Ok (t, packet, signed)
+          else
+            Ok (t, packet, false)
```

Comment:
Note to myself and other reviewers: This scary-looking change is basically just adding a `if t.dnssec then ... else Ok (t, packet, false)`. Best viewed with word-diff.